File: java/twosum/twosum.java
package twosum;
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> indices = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int difference = target - nums[i];
            if (indices.containsKey(difference)) {
                return new int[] { indices.get(difference), i };
            }
            indices.put(nums[i], i);
        }
        return new int[] {};
    }
}

public class twosum {
    public static void main(String[] args) {
        Solution trial = new Solution();
        int[] nums = { 4, 7, 1, 8, 0 };
        int target = 15;
        int[] result = trial.twoSum(nums, target);
        for (int i = 0; i < result.length; i++) {
            System.out.println(result[i]);
        }
    }
}

File: c++/twosum/two_sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> indices; // Using unordered_map for faster lookup
        for (int i = 0; i < nums.size(); i++)
        {
            int difference = target - nums[i];
            if (indices.count(difference))
            {
                return {indices[difference], i};
            }
            indices[nums[i]] = i;
        }
        return {}; // No solution found
    }
};

int main()
{
    Solution trial;
    vector<int> nums = {4, 7, 1, 8, 0};
    int target = 11;
    vector<int> result = trial.twoSum(nums, target);
    if (!result.empty())
    {
        cout << "Indices of the two numbers that add up to the target: " << result[0] << " and " << result[1] << endl;
    }
    else
    {
        cout << "No solution found." << endl;
    }
    return 0;
}